crates/rspack_plugin_javascript_shared/src/js_parser_hook.rs

```diff
@@ -0,0 +1,160 @@
+use std::{borrow::Cow, collections::HashMap};
+
+use rspack_core::{CodeGeneratableDependency, ModuleDependency};
+use swc_core::ecma::{
+  ast,
+  visit::{Visit, VisitWith},
+};
+
+type StaticStr = Cow<'static, str>;
+
+#[derive(Debug, Default)]
+pub struct JsParserContext {
+  pub presentational_dependencies: Vec<Box<dyn CodeGeneratableDependency>>,
+  pub dependencies: Vec<Box<dyn ModuleDependency>>,
+}
+
+/// The reason why we introduce [JsParserHookDriver] instead of using plain `swc_core::ecma_visit::Visit` is
+/// we want to finish all work in a single visit pass.
+#[derive(Default)]
+pub struct JsParserHookDriver {
+  hooks: HashMap<StaticStr, Vec<Box<dyn JsParserHook>>>,
```

Comment:
What is this key for?

Reply:
https://github.com/webpack/webpack/blob/f1ef2eded2fda8136535c60813dcb0c55510a5e0/lib/APIPlugin.js#L193

Reply:
But this looks not the same to me. 

For example: https://github.com/web-infra-dev/rspack/pull/3690/files#diff-8c1f9a7201ddef2c36f695634b4584ccfc786751a7c4eb2975625795b87ebe6fR83. Even after you registered the key `require.context`, you still need to write a guard here: https://github.com/web-infra-dev/rspack/pull/3690/files#diff-8c1f9a7201ddef2c36f695634b4584ccfc786751a7c4eb2975625795b87ebe6fR20.

The expected way of this would be how webpack does.

Reply:
It's not used as a guard and it doesn't affect the AST visit. Let's have a meeting for reviewing.

Reply:
It would be more convenient to use this as a guard or, personally speaking, this key does not seem necessary to me...

Reply:
No. The key has the other purpose. It's not used to express what ast node I want to visit.

Another use case of Webpack:

- https://github.com/webpack/webpack/blob/00769837bcd2bdeb105ed5786373b6c2a0c0b1e1/lib/dependencies/HarmonyImportDependencyParserPlugin.js#L30
- https://github.com/webpack/webpack/blob/00769837bcd2bdeb105ed5786373b6c2a0c0b1e1/lib/dependencies/HarmonyImportDependencyParserPlugin.js#L187

Reply:
Doesn't this mean that I want to visit identifier for `__webpack_layer__`? https://github.com/webpack/webpack/blob/f1ef2eded2fda8136535c60813dcb0c55510a5e0/lib/APIPlugin.js#L193

Reply:
Nope.

```js
parser.hooks.evaluateIdentifier
  .for("__webpack_layer__")
  .tap(PLUGIN_NAME, expr =>
```

means that I want my callback to be called when the AST node `evaluateIdentifier` is visited. The thing that the callback do is related `__webpack_layer__`, which is just a string. It's more of a description. You could write it in this way

```js
parser.hooks.evaluateIdentifier
  .for("something related to __webpack_layer__")
  .tap(PLUGIN_NAME, expr =>
```

Reply:
So you may have many callbacks that care about the same thing. The key is used to express what thing.

Reply:
I wrote a demo locally. After changing `for("__webpack_layer__")` to `for("something related to __webpack_layer__")`, the callback is never called.

Reply:
You can see here for how webpack implements this feature with the key of hook map: https://github.com/webpack/webpack/blob/6be4065ade1e252c1d8dcba4af0f43e32af1bdc1/lib/javascript/JavascriptParser.js#L3556

Reply:
Get it. Let's have a meeting. This might not fit the current design of this PR.